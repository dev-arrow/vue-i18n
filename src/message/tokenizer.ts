import {
  createScanner,
  CHAR_SP as SPACE,
  CHAR_LF as NEW_LINE
} from './scanner'
import {
  SourceLocation,
  createLocation,
  createPosition,
  Position
} from './location'

export const enum TokenTypes {
  Text, // 0
  Pipe,
  BraceLeft,
  BraceRight,
  Modulo,
  Named, // 5
  List,
  LinkedAlias,
  LinkedModifier,
  LinkedDelimiter,
  ParenLeft, // 10
  ParenRight,
  EOF
}

const enum TokenChars {
  Pipe = '|',
  BraceLeft = '{',
  BraceRight = '}',
  Modulo = '%',
  LinkedAlias = '@',
  LinkedModifier = '.',
  LinkedDelimiter = ':',
  ParenLeft = '(',
  ParenRight = ')'
}

// TODO: should be move to utils
const EOF = undefined

export type Token = {
  type: TokenTypes
  value?: string | number
  loc?: SourceLocation
}

export type Tokenizer = Readonly<{
  currentPosition: () => Position,
  nextToken: () => Token
}>

export function createTokenizer (source: string): Tokenizer {
  const _source = source
  const _scnr = createScanner(source)

  const currentPosition = (): Position => {
    return createPosition(_scnr.line(), _scnr.column(), _scnr.index())
  }

  let _currentType = TokenTypes.EOF
  let _prevType = -1
  let _currentValue: string | number | undefined | null = null
  let _startLoc = currentPosition()
  let _endLoc = _startLoc
  let _lastStartLoc: Position | null = null
  let _lastEndLoc: Position | null = null

  const getToken = (type: TokenTypes, value?: string | number): Token => {
    _endLoc = currentPosition()
    _prevType = _currentType
    _currentType = type
    _currentValue = value
    return {
      type,
      value,
      loc: createLocation(_startLoc, _endLoc)
    }
  }

  const peekSpaces = (): void => {
    while (_scnr.currentPeek() === SPACE || _scnr.currentPeek() === NEW_LINE) {
      _scnr.peek()
    }
  }

  const skipSpaces = (): void => {
    peekSpaces()
    _scnr.skipToPeek()
  }

  const isCharIdStart = (ch: string): boolean => {
    if (!ch) { return false }
    const cc = ch.charCodeAt(0)
    return (cc >= 97 && cc <= 122) || // a-z
           (cc >= 65 && cc <= 90) // A-Z
  }

  const isNumberStart = (ch: string): boolean => {
    if (!ch) {
      return false
    }
    const cc = ch.charCodeAt(0)
    return cc >= 48 && cc <= 57 // 0-9
  }

  const isNamedIdentifier = (): boolean => {
    if (_currentType !== TokenTypes.BraceLeft) { return false }
    peekSpaces()
    const ret = isCharIdStart(_scnr.currentPeek())
    _scnr.resetPeek()
    return ret
  }

  const isListIdentifier = (): boolean => {
    if (_currentType !== TokenTypes.BraceLeft) { return false }
    peekSpaces()
    const ch = _scnr.currentPeek() === '-'
      ? _scnr.peek()
      : _scnr.currentPeek()
    const ret = isNumberStart(ch)
    _scnr.resetPeek()
    return ret
  }

  const isTextStart = (): boolean => {
    if (_currentType === TokenTypes.BraceLeft ||
        _currentType === TokenTypes.Pipe ||
        _currentType === TokenTypes.LinkedAlias) {
      return false
    }

    const fn = (hasSpace = false): boolean => {
      const ch = _scnr.currentPeek()
      if ((ch === TokenChars.BraceLeft ||
        ch === TokenChars.Modulo ||
        ch === TokenChars.LinkedAlias || !ch)) {
        return hasSpace
      } else if (ch === TokenChars.Pipe) {
        return false
      } else if (ch === SPACE) {
        _scnr.peek()
        return fn(true)
      } else if (ch === NEW_LINE) {
        _scnr.peek()
        return fn(true)
      } else {
        return true
      }
    }

    const ret = fn()
    _scnr.resetPeek()
    return ret
  }

  const takeChar = (fn: Function): string | undefined | null => {
    const ch = _scnr.currentChar()
    if (ch === EOF) { return EOF }
    if (fn(ch)) {
      _scnr.next()
      return ch
    }
    return null
  }

  const takeNamedIdentifierChar = (): string | undefined | null => {
    const closure = (ch: string) => {
      const cc = ch.charCodeAt(0)
      return ((cc >= 97 && cc <= 122) || // a-z
              (cc >= 65 && cc <= 90) || // A-Z
              (cc >= 48 && cc <= 57) || // 0-9
               cc === 95 || cc === 36) // _ $
    }
    return takeChar(closure)
  }

  const takeDigit = (): string | undefined | null => {
    const closure = (ch: string) => {
      const cc = ch.charCodeAt(0)
      return (cc >= 48 && cc <= 57) // 0-9
    }
    return takeChar(closure)
  }

  const getDigits = (): string => {
    let ch: string | undefined | null = ''
    let num = ''
    while ((ch = takeDigit())) {
      num += ch
    }
    if (num.length === 0) {
      // TODO: parse errror
    }
    return num
  }

  const readText = (): string => {
    const fn = (detect = false, buf: string): string => {
      const ch = _scnr.currentChar()
      if ((ch === TokenChars.BraceLeft ||
        ch === TokenChars.Modulo ||
        ch === TokenChars.LinkedAlias || !ch)) {
        return buf
      } else if (ch === TokenChars.Pipe) {
        return buf
      } else if (ch === SPACE) {
        buf += ch
        _scnr.next()
        return fn(detect, buf)
      } else if (ch === NEW_LINE) {
        buf += ch
        _scnr.next()
        return fn(detect, buf)
      } else {
        buf += ch
        _scnr.next()
        return fn(true, buf)
      }
    }
    return fn(false, '')
  }

  const readNamedIdentifier = (): string => {
    skipSpaces()
    let ch: string | undefined | null = ''
    let name = ''
    while ((ch = takeNamedIdentifierChar())) {
      name += ch
    }
    skipSpaces()
    return name
  }

  const readListIdentifier = (): number => {
    skipSpaces()
    let value = ''
    if (_scnr.currentChar() === '-') {
      _scnr.next()
      value += `-${getDigits()}`
    } else {
      value += getDigits()
    }
    skipSpaces()
    return parseInt(value, 10)
  }

  const readToken = (): Token => {
    let token = { type: TokenTypes.EOF }
    const ch = _scnr.currentChar()
    switch (ch) {
      case TokenChars.BraceLeft:
        _scnr.next()
        token = getToken(TokenTypes.BraceLeft, TokenChars.BraceLeft)
        break
      case TokenChars.BraceRight:
        _scnr.next()
        token = getToken(TokenTypes.BraceRight, TokenChars.BraceRight)
        break
      default:
        if (isTextStart()) {
          token = getToken(TokenTypes.Text, readText())
        } else if (isNamedIdentifier()) {
          token = getToken(TokenTypes.Named, readNamedIdentifier())
        } else if (isListIdentifier()) {
          token = getToken(TokenTypes.List, readListIdentifier())
        }
        break
    }
    return token
  }

  const nextToken = (): Token => {
    _lastStartLoc = _startLoc
    _lastEndLoc = _endLoc
    _startLoc = currentPosition()

    if (!_scnr.currentChar()) {
      return getToken(TokenTypes.EOF)
    }

    return readToken()
  }

  return Object.freeze({
    nextToken,
    currentPosition
  })
}

export function parse (source: string): Token[] {
  const tokens = [] as Token[]
  const tokenizer = createTokenizer(source)
  let token: Token | null = null
  do {
    token = tokenizer.nextToken()
    tokens.push(token)
  } while (token.type !== TokenTypes.EOF)
  return tokens
}
