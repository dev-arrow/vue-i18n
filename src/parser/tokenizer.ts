// import { createScanner } from './scanner'

type Tokenizer = Readonly<{
}>

export function createTokenizer (source: string): Tokenizer {
  // const _scanner = createScanner(source)

  return Object.freeze({
    // TODO
  })
}
