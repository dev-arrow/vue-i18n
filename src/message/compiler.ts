import { createParser, ResourceNode, Node, NodeTypes } from './parser'

type CodeGenContext = {
  source?: string
  code: string
  indentLevel: number
  // line: number
  // column: number
  // offset: number
  // map?: SourceMapGenerator
}

type CodeGenerator = Readonly<{
  context: () => CodeGenContext
  push: (code: string) => void
  indent: () => void
  deindent: (withoutNewLine?: boolean) => void
  newline: () => void
}>

function createCodeGenerator (source?: string): CodeGenerator {
  const _context = {
    source,
    code: '',
    indentLevel: 0
  } as CodeGenContext

  const context = (): CodeGenContext => _context

  const push = (code: string): void => {
    _context.code += code
  }

  const _newline = (n: number): void => {
    push('\n' + `  `.repeat(n))
  }

  const indent = (): void => {
    _newline(++_context.indentLevel)
  }

  const deindent = (withoutNewLine?: boolean): void => {
    if (withoutNewLine) {
      --_context.indentLevel
    } else {
      _newline(--_context.indentLevel)
    }
  }

  const newline = (): void => {
    _newline(_context.indentLevel)
  }

  return Object.freeze({
    context, push, indent, deindent, newline
  })
}

export type CompileOptions = {
  // TODO: should be defined some options
  // Generate source map?
  // - Default: false
  // sourceMap?: boolean
  // Filename for source map generation.
  // - Default: `message.intl`
  // filename?: string
}

export type CompileResult = {
  code: string
  ast: ResourceNode
  // TODO: should be implemetend sourcemap
  // map?: RawSourceMap
}

export type Compiler = Readonly<{
  compile: (source: string, options?: CompileOptions) => CompileResult
}>

export function createCompiler (): Compiler {
  const _parser = createParser()

  // transform AST
  const transform = (ast: ResourceNode, options: CompileOptions): void => {
    // TODO:
  }

  const generateNode = (generator: CodeGenerator, node: Node): void => {
    generator.push(`''`)
    /*
    switch (node.type) {
      default:
        // TODO: should be handled with error
        throw new Error(`unhandled codegen node type: ${node.type}`)
    }
    */
  }

  // generate code from AST
  const generate = (ast: ResourceNode, options: CompileOptions): string => {
    const generator = createCodeGenerator(ast.loc && ast.loc.source)
    generator.push(`function __msg__ (ctx) {`)
    generator.indent()
    generator.push('return ')
    generateNode(generator, ast)
    generator.deindent()
    generator.push(`}`)
    return generator.context().code
  }

  const compile = (source: string, options: CompileOptions = {}): CompileResult => {
    const ast = _parser.parse(source)
    transform(ast, options)
    const code = generate(ast, options)
    return { code, ast }
  }

  return Object.freeze({
    compile
  })
}
