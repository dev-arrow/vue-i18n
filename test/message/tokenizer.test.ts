import { parse } from '../../src/message/tokenizer'

test('parse', () => {
  [
    'hello world',
    'hello\nworld',
    '',
    ' hello world ',
    'hi {name} !',
    '{first} {middle}　{last}',
    'hi {  name } !',
    '{first}\n{middle}\r\n{last}',
    'hi {0} !',
    '{0} {1}　{2}',
    'hi {  -1 } !',
    '{0}\n{1}\r\n{2}'
  ].forEach(p => {
    expect(parse(p)).toMatchSnapshot(JSON.stringify(p))
  })
})
