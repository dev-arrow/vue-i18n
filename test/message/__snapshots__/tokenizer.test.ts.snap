// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse: " hello world " 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": " hello world ",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
      "source": undefined,
      "start": Object {
        "column": 14,
        "line": 1,
        "offset": 13,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "{0} {1}　{2}" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "type": 6,
    "value": 0,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "source": undefined,
      "start": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
    },
    "type": 0,
    "value": " ",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "source": undefined,
      "start": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
      "source": undefined,
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
    },
    "type": 6,
    "value": 1,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "source": undefined,
      "start": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "source": undefined,
      "start": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "type": 0,
    "value": "　",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "source": undefined,
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "source": undefined,
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
    },
    "type": 6,
    "value": 2,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "{0}\\n{1}\\r\\n{2}" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "type": 6,
    "value": 0,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "source": undefined,
      "start": Object {
        "column": 3,
        "line": 1,
        "offset": 2,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 2,
        "offset": 4,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
    },
    "type": 0,
    "value": "
",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 2,
        "offset": 5,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 2,
        "offset": 4,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 2,
        "offset": 6,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 2,
        "offset": 5,
      },
    },
    "type": 6,
    "value": 1,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 2,
        "offset": 7,
      },
      "source": undefined,
      "start": Object {
        "column": 3,
        "line": 2,
        "offset": 6,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 3,
        "offset": 9,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 2,
        "offset": 7,
      },
    },
    "type": 0,
    "value": "
",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 3,
        "offset": 10,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 3,
        "offset": 9,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 3,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 3,
        "offset": 10,
      },
    },
    "type": 6,
    "value": 2,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 3,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 3,
        "line": 3,
        "offset": 11,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 3,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 3,
        "offset": 12,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "{first} {middle}　{last}" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "type": 5,
    "value": "first",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "source": undefined,
      "start": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "source": undefined,
      "start": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "type": 0,
    "value": " ",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "source": undefined,
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
      "source": undefined,
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
    },
    "type": 5,
    "value": "middle",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
      "source": undefined,
      "start": Object {
        "column": 16,
        "line": 1,
        "offset": 15,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 18,
        "line": 1,
        "offset": 17,
      },
      "source": undefined,
      "start": Object {
        "column": 17,
        "line": 1,
        "offset": 16,
      },
    },
    "type": 0,
    "value": "　",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "source": undefined,
      "start": Object {
        "column": 18,
        "line": 1,
        "offset": 17,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 23,
        "line": 1,
        "offset": 22,
      },
      "source": undefined,
      "start": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
    },
    "type": 5,
    "value": "last",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 24,
        "line": 1,
        "offset": 23,
      },
      "source": undefined,
      "start": Object {
        "column": 23,
        "line": 1,
        "offset": 22,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 24,
        "line": 1,
        "offset": 23,
      },
      "source": undefined,
      "start": Object {
        "column": 24,
        "line": 1,
        "offset": 23,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "{first}\\n{middle}\\r\\n{last}" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 1,
        "offset": 1,
      },
    },
    "type": 5,
    "value": "first",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
      "source": undefined,
      "start": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 2,
        "offset": 8,
      },
      "source": undefined,
      "start": Object {
        "column": 8,
        "line": 1,
        "offset": 7,
      },
    },
    "type": 0,
    "value": "
",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 2,
        "offset": 9,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 2,
        "offset": 8,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 8,
        "line": 2,
        "offset": 15,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 2,
        "offset": 9,
      },
    },
    "type": 5,
    "value": "middle",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 2,
        "offset": 16,
      },
      "source": undefined,
      "start": Object {
        "column": 8,
        "line": 2,
        "offset": 15,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 3,
        "offset": 18,
      },
      "source": undefined,
      "start": Object {
        "column": 9,
        "line": 2,
        "offset": 16,
      },
    },
    "type": 0,
    "value": "
",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 2,
        "line": 3,
        "offset": 19,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 3,
        "offset": 18,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 3,
        "offset": 23,
      },
      "source": undefined,
      "start": Object {
        "column": 2,
        "line": 3,
        "offset": 19,
      },
    },
    "type": 5,
    "value": "last",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 3,
        "offset": 24,
      },
      "source": undefined,
      "start": Object {
        "column": 6,
        "line": 3,
        "offset": 23,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 3,
        "offset": 24,
      },
      "source": undefined,
      "start": Object {
        "column": 7,
        "line": 3,
        "offset": 24,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "hello world" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hello world",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "hello\\nworld" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 2,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hello
world",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 2,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 6,
        "line": 2,
        "offset": 11,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "hi {  -1 } !" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hi ",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "source": undefined,
      "start": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
    },
    "type": 6,
    "value": -1,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
      "source": undefined,
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 11,
        "line": 1,
        "offset": 10,
      },
    },
    "type": 0,
    "value": " !",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "hi {  name } !" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hi ",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
    },
    "type": 5,
    "value": "name",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "source": undefined,
      "start": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
    },
    "type": 0,
    "value": " !",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
      "source": undefined,
      "start": Object {
        "column": 15,
        "line": 1,
        "offset": 14,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "hi {0} !" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hi ",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
      "source": undefined,
      "start": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
    },
    "type": 6,
    "value": 0,
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
      "source": undefined,
      "start": Object {
        "column": 6,
        "line": 1,
        "offset": 5,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "source": undefined,
      "start": Object {
        "column": 7,
        "line": 1,
        "offset": 6,
      },
    },
    "type": 0,
    "value": " !",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "source": undefined,
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;

exports[`parse: "hi {name} !" 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hi ",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
      "source": undefined,
      "start": Object {
        "column": 4,
        "line": 1,
        "offset": 3,
      },
    },
    "type": 2,
    "value": "{",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
      "source": undefined,
      "start": Object {
        "column": 5,
        "line": 1,
        "offset": 4,
      },
    },
    "type": 5,
    "value": "name",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
      "source": undefined,
      "start": Object {
        "column": 9,
        "line": 1,
        "offset": 8,
      },
    },
    "type": 3,
    "value": "}",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 10,
        "line": 1,
        "offset": 9,
      },
    },
    "type": 0,
    "value": " !",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "source": undefined,
      "start": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
    },
    "type": 12,
    "value": undefined,
  },
]
`;
